File: ConvexHull.m
function [Y, WtildeMatrix,Indices]=ConvexHull(X,K,lambda,gamma)

%Convex Hull stratification algorithm detailed in article
%L. Ziegelmeier, M. Kirby, C. Peterson, "Stratifying High-Dimensional Data 
%Based on Proximity to the Convex Hull Boundary," SIAM Review, Vol. 59, No.
%2, pp. 346-365, 2017.

%Code by Lori Ziegelmeier, note this code has not been optimized

%Input:
% X = Dxp data set X
% K = the number of nearest neighbors
% lambda = parameter on the l_1 convexity term 
% gamma = parameter on the l_2 uniformity term

%Outputs: 
% Y = candidates for vertices of the convex hull
% WtildeMatrix = the Kxp matrix corresponding to weight vectors associated 
%                to each point (note weight vectors with negative entries 
%                are candidate vertices)
% Indices = indices of the original data set X with negative entries in the
%           weight vector


%warning('off','all');


S=size(X);
p = S(2) %Number of points
D= S(1) %Dimensionality


%%%Computing the nearest neighbors
tic;
disp('Computing distance matrix')
SumSqX=sum(X.^2,1);  
SqDist=repmat(SumSqX,p,1)+repmat(SumSqX',1,p)-2*X'*X; %Computes the distance squared between all points using the equation ||x-y|| = sqrt(||x||+||y||-2< x,y>)
Distance =sqrt(SqDist); %finding the distance between all of the points
[SortDist, Index] = sort(Distance); %Sorts each row of the distance matrix
disp('Done computing distance matrix')
%Must use when there are repeated points
NeighborInd=zeros(K,p);
for j=1:p
    DiffPoints=find(SortDist(:,j)); %finds all nonzero distances
    NeighborInd(:,j)=Index(DiffPoints(1):DiffPoints(1)+K-1,j);
end
NeighborInd;
disp('Done Computing Nearest Neighbors')
toc


%Creating a data cube containing the matrix of neighborhoods for each
%point Xi
N=zeros(D,K,p);
for i=1:p
    Ni=zeros(D,K);
    for t=1:K
        Ni(:,t)=X(:,NeighborInd(t,i));
    end
    N(:,:,i)=Ni;
end
N;

%Computing the Reconstruction Weights by solving a quadratic program
%Decision variables are the nonzero entries of W, wtilde, rewritten as
%wtildeplus and wtildeminus
%Interested in the negative entries, as should indicate points on boundary
%or vertices

%Forming the H matrix and the f vector
disp('Forming the Quadratic Program')
tic
H=zeros(2*K,2*K,p);
f=zeros(2*K,p);
for i=1:p
    Ni=N(:,:,i); %Forming a matrix of all neighbors of the point Xi
    Hhat=zeros(K,K);
    ftilde=zeros(K,1);
    fhat=zeros(K,1);
    Xi=X(:,i);
    for j=1:K
        Xj=Ni(:,j);
        ftilde(j)=-2*Xi'*Xj+lambda;
        fhat(j)=2*Xi'*Xj+lambda;
        for l=1:K
            Xl=Ni(:,l);
            Hhat(j,l)=Xj'*Xl;
        end
    end
    Htilde=[Hhat -Hhat; -Hhat Hhat];
    H(:,:,i)=Htilde;
    f(:,i)=[ftilde; fhat];
end
toc
disp('Done forming quadratic program')

disp('Solving the Quadratic Program')
tic
Aeq=[ones(1,K) -ones(1,K)];
b=1;
WtildeMatrix=zeros(K,p);
fval=0;
Y=[];
Indices=[];
for i=1:p
    i;
    Htilde=H(:,:,i);
    ftilde=f(:,i);
    %[x,fval,exitflag,output,lambda] = quadprog(2*(Htilde+gamma*[eye(K) -eye(K); -eye(K) eye(K)]),ftilde,-eye(2*K),zeros(2*K,1),Aeq,b,'options','interior-point-convex');
    
    [output, x,lambdaout, nu] = PDIPAQuad(2*(Htilde+gamma*[eye(K) -eye(K); -eye(K) eye(K)]),ftilde,Aeq,b,10^-10);
    
    Wtildeplus=x(1:K);
    Wtildeminus=x(K+1:2*K);
    Wtilde=Wtildeplus-Wtildeminus;
    Wtilde=Wtilde/sum(Wtilde);
    WtildeMatrix(:,i)=Wtilde;
    if isempty(find(Wtilde<0))==0 %determining if the weight vector has negative entries
        Y=[Y X(:,i)];
        Indices=[Indices i];
    end
    
    Verify=norm(X(:,i)-N(:,:,i)*Wtilde)^2; %Make sure that each point reconstructed perfectly
   
end
toc
disp('Done solving the Quadratic Program')

File: convex_hull.py
def convex_hull(X,K,lam,gamma,**kwargs):
    '''
    A transliteration of Lor Ziegelmeier's ConvexHull.m code.
    Primary difference is that **the data matrix X should be
    arranged by rows**, as is typical in python/sklearn/calcom.

    The convex hull stratification algorithm detailed in
    the article

        L. Ziegelmeier, M. Kirby, C. Peterson,
        "Stratifying High-Dimensional Data Based on proximity
        to the Convex Hull Boundary", SIAM Review, Vol. 59, No. 2, pp.346-365, 2017

    Inputs:
        X : p-by-D dataset X
        K : number of nearest neighbors
        lam : parameter on the l_1 convexity term
        gamma : parameter on the l_2 uniformity term

    Outputs:
        Y : candidates for vertices of the convex hull
        WtildeMatrix : the K-by-p matrix corresponding to weight
            vectors associated to each point (note weight vectors
            with negative entries are candidate vertices)
        Indices : indices of the original data set X with negative
            entries in the weight vector

    Optional inputs:
        verbosity : integer. Set greater than zero for print statements
            (default: 0)
        metric : string, indicating the metric to be used. This is
            passed on to sklearn.metrics.pairwise_distances();
            see that documentation for the full list of options.
            Notable options include 'cosine', 'l1', and 'l2'.
            (default: 'euclidean')
    '''
    import numpy as np
    from sklearn import metrics

    import opt  # contains quadratic program optimization routine

    verbosity = kwargs.get('verbosity',0)
    metric = kwargs.get('metric', 'euclidean')

    p,D = np.shape(X)
    if verbosity>0: print('Computing distance matrix')

    Distance = metrics.pairwise_distances(X, metric=metric)
    # We want a sorted ilst of nearest neighbors for each
    # point.
    nneighbors = [np.argsort(row) for row in Distance]

    if verbosity>0: print('Done computing distance matrix')

    NeighborInd = np.zeros((p,K), dtype=int)
    for j in range(p):
        DiffPoints = np.where( Distance[j]>0. )[0]
        idx = 0
        for k in nneighbors[j]:
            if Distance[j][k]>0:
                NeighborInd[j][idx] = k
                idx += 1
            if idx==K:
                break
        #
    #
    if verbosity>0: print('Done computing nearest neighbors')

    # Creating data cube containing the matrix of
    # neighborhoods for each point X_i
    #
    N = np.zeros( (p,K,D) )
    for i in range(p):
        N[i] = X[NeighborInd[i]]
    #

    # Computing the reconstruction weights by solving
    # a quadratic program.
    # Decision variables are the nonzero entries of W,
    # wtilde, rewritten as wtildeplus and wtildeminus
    # Interested in the negative entries, as should indicate
    # points on boundary or vertices.
    if verbosity>0: print('Forming the quadratic program')
    H = np.zeros((p, 2*K, 2*K))
    f = np.zeros((p, 2*K))

    # to be used with the Kronecker product
    tiling_array = np.array([[1,-1],[-1,1]])

    for i in range(p):
        # Not the best naming scheme, but whatever
        doti = np.dot(X[i],N[i].T)      # dot products of point of interest onto its neighbors
        NNgram = np.dot(N[i], N[i].T)   # dot products between all neighbors (Gram matrix)

        ftilde = -2*doti + lam
        fhat = 2*doti + lam

        Hhat = NNgram
        Htilde = np.kron(tiling_array, Hhat)

        H[i] = Htilde
        f[i] = np.concatenate( [ftilde,fhat] )
    #

    if verbosity>0: print('Done forming quadratic program')
    if verbosity>0: print('Solving quadratic program')


    Aeq = np.kron([1,-1], np.ones(K))
    Aeq.shape = (1,len(Aeq))
    b=np.array([1]);
    WtildeMatrix = np.zeros((p,K))
    fval=0;
    Y=[];
    Indices=[];

#    import pdb
#    pdb.set_trace()

    for i in range(p):
        Htilde = H[i]
        ftilde = f[i]

        Q = 2*(Htilde + gamma*np.kron(tiling_array,np.eye(K)))
        x = opt.PDIPAQuad(Q, ftilde, Aeq, b, 10**-8)

        # Reconstruct weights
        Wtilde = x[:K] - x[K:2*K]
        Wtilde /= sum(Wtilde)
        WtildeMatrix[i] = Wtilde

        # Determining if the weight vector has negative entries
        loc = np.where(Wtilde<0.)[0]

        if len(loc)==0:
            Y.append(X[i])
            Indices.append(i)
        else:
            # print(loc,i)
            pass
        #

        # Checking if each point recnostructed perfectly
        # Verify = np.linalg.norm(X[i] - N[i])**2
    #

    if verbosity>0: print('Done solving the quadratic program')

    return Y,WtildeMatrix,Indices
#

if __name__=="__main__":
    import numpy as np
    from matplotlib import pyplot

    th = np.linspace(0,2*np.pi,51)
    x = np.cos(th) + 0.2*np.random.randn(51)
    y = np.sin(th) + 0.2*np.random.randn(51)
    X = np.array([[xv,yv] for xv,yv in zip(x,y)])

    Y,Wt,Indices = convex_hull(X,4,1,1)

    fig,ax = pyplot.subplots(1,1)
    ax.scatter(x,y)
    ax.axis('square')
    fig.show()
    pyplot.ion()
#
